### ISSUE TITLE
FPC built-in functions redefined in rpi_hal with different parameters

### ISSUE THREAD

**Issue by kaimex** [rudiratlos/rpi-hal#2]
Within rpi_hal some FPC functions have been redefined.
E.g. hexstr() , upper(), lower()  and others.
hexstr() uses a different number of parameters.
People who try to use FPC's hexstr() function will receive an errormessage, complaining about to many parameters.
Those who want to use FPC's built-in functions will have to rename all redefined functions in rpi_hal.

Kaimex

**Comment by rudiratlos** [comment]
you are right with upper/lower. these functions I've just imprted from my turbo pascal libraries form 30 yeras ago. HexStr is for strings. free pascal offers HexStr only for numbers (as I know).

**Comment by kaimex** [comment]
HexStr: may be, but it is not a good idea to block access to a FPC built-in function by defining a function for a different purpose with the same name.

**Comment by rudiratlos** [comment]
which functions? Upper/lower is not in free pascal dialect, as I know

**Comment by kaimex** [comment]
HexStr is defined in FPC to have 2 parameters, an integer type and  the number of hex digits to produce. You have defined HexStr as a function with only one parameter, a string. It took me some time to find out, why Lazarus complained my correct use of the FPC function. FPC has the two functions UpperCase and LowerCase, so there is no need to define the same function as Upper and Lower in your library.

**Comment by rudiratlos** [comment]
HexStr can have 1 or 2 parameter ( pls. see https://www.freepascal.org/docs-html/rtl/system/hexstr.html ) and look for pointer as parameter. I've just extended HexStr for string as parameter.
Upper/lower function, comes from legacy. I have a lot of old personal pascal sources, using these functions and other. I'm really appreciate your comments. I want to extend and improve rpi-hal.

**Comment by kaimex** [comment]
I was using Lazarus as the frontend for FPC. My modfied version of your test-program for rpi_hal V4.5 used HexStr(word_parameter,8) as defined in the FPC documentation to output a register value of the ADS1115 ADC in hex. It simply did not compile with the error message "not enough parameters" or similar, because your version of HexStr blocked access to the original version.  Only after I renamed your HexStr to HexStr2 the program was successfully compiled.
If you want to improve rpi_hal, I would give the advice: remove everything which is not really necessary. It is really hard for somebody else, to figure out what's going on in it, what is really essential and what is luxury. I would also remove the ham/morse stuff and put that into another related library. A lean modular approach to all your ideas would make things more easily servicable/manageable/modifyable.
I would also suspect, that you won't make many friends by changing the parameter list e.g. of the I2C-function from version to version. By the way, their description in the read-me of version 4.5 did not reflect the actual parameters. If you really have a good reason (possibly a major advantage) to change parameters, you should explain why. Otherwise it will appear just arbitrary.
A library is only as valuable as the documentation of its (or at least of the essential) functions and procedures. Most people won't ever use functions buried in the depths of the library and not mentioned in an interface list, because they don't have the time to look into the code or even to try to understand it.

**Comment by rudiratlos** [comment]
just put overload to HexStr with string parameter. now it compiles with one parameter.
Thanx for your feedback

**Comment by kaimex** [comment]
I prefer to use different names for different functions. Overloading reduces the the chances of the compiler to find programming errors.
